File: 530-minimum-absolute-difference-in-bst/solution.py
from typing import Optional
from collections import deque
import math


class TreeNode:
  def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
    work_deque: deque[TreeNode] = deque()
    curr = root
    last_visited: Optional[TreeNode] = None
    min_diff = math.inf
    while curr is not None or work_deque:
      if curr is not None:
        work_deque.append(curr)
        curr = curr.left
        continue

      node = work_deque.pop()
      curr = node.right
      if last_visited is None:
        last_visited = node
        continue

      min_diff = min(min_diff, abs(node.val - last_visited.val))
      last_visited = node

    return min_diff

File: 530-minimum-absolute-difference-in-bst/solution.cpp
#include <deque>
#include <limits>
#include <algorithm>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  int getMinimumDifference(TreeNode *root)
  {
    std::deque<const TreeNode *> work_deque;
    const TreeNode *curr = root;
    const TreeNode *last_visited = nullptr;
    int min_diff = std::numeric_limits<int>::max();
    while (curr != nullptr || !work_deque.empty())
    {
      if (curr != nullptr)
      {
        work_deque.emplace_back(curr);
        curr = curr->left;
        continue;
      }

      const TreeNode *const node = work_deque.back();
      work_deque.pop_back();
      curr = node->right;
      if (last_visited == nullptr)
      {
        last_visited = node;
        continue;
      }

      min_diff = std::min(min_diff, abs(node->val - last_visited->val));
      last_visited = node;
    }

    return min_diff;
  }
};